x/gasless/keeper/fee_helper.go

```diff
@@ -0,0 +1,238 @@
+	gasTank, found := k.GetGasTank(ctx, gtid)
+	// there is no gas tank with given id, likely impossible to happen
+	// exists only as aditional check.
+	if !found {
+		return gasTank, false, sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "gas tank not found")
+	}
+
+	// gas tank is not active and cannot be used as fee source
+	if !gasTank.IsActive {
+		return gasTank, false, sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "gas tank not active")
+	}
+
+	// fee denom does not match between gas tank and asked fee
+	if fee.Denom != gasTank.FeeDenom {
+		return gasTank, false, sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "denom mismatch between tank and asked fee")
+	}
+
+	// asked fee amount is more than the allowed fee usage for tx.
+	if fee.Amount.GT(gasTank.MaxFeeUsagePerTx) {
+		return gasTank, false, sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "fee amount more than allowed limit")
+	}
+
+	// insufficient reserve in the gas tank to fulfill the transaction fee
+	gasTankReserveBalance := k.bankKeeper.GetBalance(ctx, gasTank.GetGasTankReserveAddress(), gasTank.FeeDenom)
+	if gasTankReserveBalance.IsLT(fee) {
+		return gasTank, false, sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "funds insufficient in gas reserve tank")
+	}
+
+	found = false
+	consumptionIndex := 0
+	for index, consumption := range consumer.Consumptions {
+		if consumption.GasTankId == gasTank.Id {
+			consumptionIndex = index
+			found = true
+		}
+	}
+	// no need to check the consumption usage since there is no key available with given gas tank id
+	// i.e the consumer has never used this gas reserve before and the first time visitor for the given gas tank
+	if !found {
+		return gasTank, true, nil
+	}
+
+	consumptionDetails := consumer.Consumptions[consumptionIndex]
+
+	// consumer is blocked by the gas tank
+	if consumptionDetails.IsBlocked {
+		return gasTank, false, sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "blocked by gas tank")
+	}
+
+	// consumer exhausted the transaction count limit, hence not eligible with given gas tank
+	if consumptionDetails.TotalTxsMade >= consumptionDetails.TotalTxsAllowed {
+		return gasTank, false, sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "exhausted tx limit")
+	}
+
+	// if total fees consumed by the consumer is more than or equal to the allowed consumption
+	// i.e consumer has exhausted its fee limit and hence is not eligible for the given tank
+	totalFeeConsumption := consumptionDetails.TotalFeesConsumed.Add(fee.Amount)
+	if totalFeeConsumption.GTE(consumptionDetails.TotalFeeConsumptionAllowed) {
+		return gasTank, false, sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "exhausted total fee usage or pending fee limit insufficient for tx")
+	}
+
+	return gasTank, true, nil
+}
+
+func (k Keeper) GetFeeSource(ctx sdk.Context, sdkTx sdk.Tx, originalFeePayer sdk.AccAddress, fees sdk.Coins) sdk.AccAddress {
+	if len(sdkTx.GetMsgs()) > 1 {
```

Comment:
fee mentioned in the tx is for entire tx and is not for individual message.
If there are multiple message types or contract execute calls, even though the fee could be free for few of them we cannot allow the whole tx to go through the tank, since it can have multiple non-whitelisted messageTypes or Contract Address.